include/proxy-wasm/context.h

```diff
@@ -0,0 +1,649 @@
+  // The VM Context used for calling "malloc" has an id_ == 0.
+  bool isVmContext() { return id_ == 0; }
+  // Root Contexts have the VM Context as a parent.
+  bool isRootContext() { return root_context_id_ == 0; }
+  ContextBase *root_context() { return root_context_; }
+  string_view root_id() const { return plugin_->root_id_; }
+  string_view log_prefix() const { return plugin_->log_prefix(); }
+  WasmVm *wasmVm() const;
+
+  /**
+   * Calls into the VM.
+   * These are implemented by the proxy-independent host code. They are virtual to support some
+   * types of testing.
+   */
+
+  // Root Context
+
+  /**
+   * Call on a Root Context when a plugin is configured or reconfigured dynamically.
+   * @param plugin is the plugin which is being configured or reconfigured.
+   */
+  virtual bool onConfigure(std::shared_ptr<PluginBase> plugin);
+
+  // Context
+
+  /**
+   * Call on a host Context to create a corresponding Context in the VM.  Note:
+   * onNetworkNewConnection and onRequestHeaders() call onCreate().
+   * @param parent_context_id is the parent Context id for the context being created.  For a Network
+   * stream Context, this will be a Root Context id.
+   */
+  virtual void onCreate(uint32_t parent_context_id);
+
+  /**
+   * Network - see network_interface.h
+   */
+  ProxyAction onNetworkNewConnection() override;
+  ProxyAction onDownstreamData(uint32_t data_length, bool end_of_stream) override;
+  ProxyAction onUpstreamData(uint32_t data_length, bool end_of_stream) override;
+  using CloseType = NetworkInterface::CloseType;
+  void onDownstreamConnectionClose(CloseType close_type) override;
+  void onUpstreamConnectionClose(CloseType close_type) override;
+
+  /*
+   * HTTP - see http_interface.h
+   */
+  ProxyAction onRequestHeaders() override;
+  ProxyAction onRequestBody(uint32_t body_buffer_length, bool end_of_stream) override;
+  ProxyAction onRequestTrailers() override;
+  ProxyAction onRequestMetadata() override;
+  ProxyAction onResponseHeaders() override;
+  ProxyAction onResponseBody(uint32_t body_buffer_length, bool end_of_stream) override;
+  ProxyAction onResponseTrailers() override;
+  ProxyAction onResponseMetadata() override;
+
+  // Root Context async events
+
+  /**
+   * Called on a Root Context when a response arrives for an outstanding httpCall().
+   * @param token is the token returned by the corresponding httpCall().
+   * @param headers are the number of headers in the response.
+   * @param body_size is the size in bytes of the response body.
+   * @param trailers is the number of trailers in the response.
+   */
+  virtual void onHttpCallResponse(uint32_t token, uint32_t headers, uint32_t body_size,
+                                  uint32_t trailers);
+
+  /**
+   * Called on a Root Context to indicate that an Inter-VM shared queue message has arrived.
+   * @token is the token returned by registerSharedQueue().
+   */
+  virtual void onQueueReady(uint32_t token);
```

Comment:
Is there only ever one kind of token? or are there different tokens for shared messages, HTTP calls etc? Should these have wrapping Struct types? Should they be uint64_t if it's a global namespace?

Reply:
These are different types of tokens.  Sadly C++ doesn't have a native way to form by-name primitive types.  Google has wrappers for that but I don't know of an OSS or STD library.  Rather than take a dependency we are doing it with raw uints.   Perhaps naming and documentation or should we add a struct wrappers.  Personally I feel like the latter  is overkill, but I can be convinced if others disagree.

Reply:
I'm fine either way. You might consider some type of `using` type alias just for readability?

Reply:
SGTM

Reply:
Yeah, I'm happy with `using` just for the readability win.